samples/msal-node-samples/refresh-token/adalCustomCache.js

```diff
@@ -0,0 +1,114 @@
+var _ = require('underscore');
+var fs = require('fs');
+
+/**
+ * Constructs a new token cache on disk.
+ * This effectively implements the ADAL TokenCache interface.
+ * @constructor
+ */
+function DiskCache(cacheLocation) {
+    this._cacheLocation = cacheLocation;
+    this._entries = [];
+}
+
+/**
+ * Removes a collection of entries from the cache in a single batch operation.
+ * @param {Array} entries An array of cache entries to remove.
+ * @param {Function} callback This function is called when the operation is complete. Any error is provided as the first parameter.
+ */
+DiskCache.prototype.remove = function (entries, callback) {
+    const cachedEntries = _readCache.call(this);
+
+    var updatedEntries = _.filter(cachedEntries, function (element) {
+        if (_.findWhere(entries, element)) {
+            return false;
+        }
+        return true;
+    });
+
+    _writeCache.call(this, updatedEntries);
+    callback();
+};
+
+/**
+ * Adds a collection of entries to the cache in a single batch operation.
+ * @param {Array} entries  An array of entries to add to the cache.
+ * @param {Function} callback This function is called when the operation is complete. Any error is provided as the first parameter.
+ */
+DiskCache.prototype.add = function (entries, callback) {
+    const cachedEntries = _readCache.call(this);
+
+    // Remove any entries that are duplicates of the existing
+    // cache elements.
+    _.each(cachedEntries, function (element) {
+        _.each(entries, function (addElement, index) {
+            if (entriesHaveEqualHashKeys(element, addElement)) {
+                entries[index] = null;
+            }
+        });
+    });
+
+    // Add the new entries to the end of the cache.
+    entries = _.compact(entries);
+
+    for (var i = 0; i < entries.length; i++) {
+        cachedEntries.push(entries[i]);
+    }
```

Comment:
use a .forEach here

Reply:
that would be the better choice semantically, but `forEach` is also ES6 -so just a stylistic choice (as well as this code was adopted from ADAL-Node [in-memory token cache](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/maintenance/adal-node/lib/memory-cache.js), but I'm happy to reconsider!)